Sources/NIOSSL/SSLContext.swift

```diff
@@ -112,6 +112,101 @@ private func alpnCallback(ssl: OpaquePointer?,
     return SSL_TLSEXT_ERR_OK
 }
 
+/// PSK Callback for the server side context.
+private func serverPSKCallback(ssl: OpaquePointer?,
+                                 identity: UnsafePointer<CChar>?,
+                                 psk: UnsafeMutablePointer<UInt8>?,
+                                 max_psk_len: UInt32) -> UInt32 {
+    
+    guard let ssl = ssl else { return 0 }
+    
+    // Initial implementation only supports TLS 1.2 due API support exposed from BoringSSL.
+    // TODO (meaton) add TLS 1.3 support when available.
+    
+    // We want to take the SSL pointer and extract the parent Swift object.
+    let parentCtx = CNIOBoringSSL_SSL_get_SSL_CTX(ssl)!
+    let parentPtr = CNIOBoringSSLShims_SSL_CTX_get_app_data(parentCtx)!
+    let parentSwiftContext: NIOSSLContext = Unmanaged.fromOpaque(parentPtr).takeUnretainedValue()
+    
+    guard let serverCallback = parentSwiftContext.pskServerConfigurationCallback,
+          let outputIdentity = identity, // Output PSK identity.
+          let strOutputIdentity = String(validatingUTF8: outputIdentity), // Map identity to str.
+          let outputPSK = psk // Output PSK key.
+          else {
+        return 0
+    }
+    
+    // Get PSK client identity hint based on the SSL connection
+    let hint = CNIOBoringSSL_SSL_get_psk_identity_hint(ssl)
+    guard let clientHint = hint,
+          let derivedHint = String(validatingUTF8: clientHint) else {
+        return 0
+    }
+    // Take the hint and the possible identity and pass them down to the callback to get associated PSK from callback
+    guard let contextPSKIdentityCallback = try? serverCallback(derivedHint, strOutputIdentity) else {
+        return 0
+    }
+    let serverPSK = contextPSKIdentityCallback.key // From the callback
+    
+    // Make sure the key is returned by the callback and it is of proper length, otherwise, fail.
+    if serverPSK.isEmpty || serverPSK.count > max_psk_len {
+        return 0
+    }
+    var pskBuffer: [UInt8] = []
+    serverPSK.backing.withUnsafeBytes {
+        pskBuffer.append(contentsOf: $0)
+    }
+    memcpy(outputPSK, pskBuffer, pskBuffer.count)
```

Comment:
`withUnsafeBytes` is on the `backing` of `NIOSSLSecureBytes`.  The reason I copied this into a byte array first is because if I do:  
```swift
 memcpy(outputPSK, serverPSK.backing.withUnsafeBytes, pskBuffer.count)
```
I get:
```
Cannot convert value of type '((UnsafeRawBufferPointer) throws -> T) throws -> T' to expected argument type 'UnsafeRawPointer?'
Generic parameter 'T' could not be inferred
```

Reply:
Sorry, what I meant is:

```swift
serverPSK.withUnsafeBytes {
    outputPSK, $0.baseAddress, $0.count
}
```

Reply:
Addressed in 732e883